bundles/org.openhab.binding.broadlinkthermostat/src/main/java/org/openhab/binding/broadlinkthermostat/internal/handler/FloureonThermostatHandler.java

```diff
@@ -0,0 +1,260 @@
+                } else if (SENSOR_EXTERNAL.equals(command.toFullString())) {
+                    floureonDevice.setMode(statusInfo.getAutoMode(), statusInfo.getLoopMode(), SensorControl.EXTERNAL);
+                } else {
+                    floureonDevice.setMode(statusInfo.getAutoMode(), statusInfo.getLoopMode(),
+                            SensorControl.INTERNAL_TEMP_EXTERNAL_LIMIT);
+                }
+            } catch (Exception e) {
+                logger.warn("Error while trying to set sensor mode {}: {}", command, e.getMessage());
+            }
+        } else {
+            logger.warn("Channel {} does not support command {}", channelUID, command);
+        }
+    }
+
+    private void handleRemoteLockCommand(ChannelUID channelUID, Command command) {
+        FloureonDevice floureonDevice = this.floureonDevice;
+        if (command instanceof OnOffType && floureonDevice != null) {
+            try {
+                floureonDevice.setLock(command == OnOffType.ON);
+            } catch (Exception e) {
+                logger.warn("Error while setting remote lock of {} to {}: {}", thing.getUID(), command, e.getMessage());
+            }
+        } else {
+            logger.warn("Channel {} does not support command {}", channelUID, command);
+        }
+    }
+
+    private void handleSetTimeCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof DateTimeType) {
+            ZonedDateTime zonedDateTime = ((DateTimeType) command).getZonedDateTime();
+            try {
+                new SetTimeCommand(tob(zonedDateTime.getHour()), tob(zonedDateTime.getMinute()),
+                        tob(zonedDateTime.getSecond()), tob(zonedDateTime.getDayOfWeek().getValue()))
+                                .execute(floureonDevice);
+            } catch (Exception e) {
+                logger.warn("Error while setting time of {} to {}: {}", thing.getUID(), command, e.getMessage());
+            }
+        } else {
+            logger.warn("Channel {} does not support command {}", channelUID, command);
+        }
+    }
+
+    @Override
+    protected void refreshData() {
+        if (ThingStatus.ONLINE != thing.getStatus()) {
+            return;
+        }
+
+        FloureonDevice floureonDevice = this.floureonDevice;
+        if (floureonDevice != null) {
+            try {
+                AdvancedStatusInfo advancedStatusInfo = floureonDevice.getAdvancedStatus();
+                if (advancedStatusInfo == null) {
+                    logger.warn("Device {} did not return any data. Trying to reauthenticate...", thing.getUID());
+                    authenticate();
+                    advancedStatusInfo = floureonDevice.getAdvancedStatus();
+                }
+                if (advancedStatusInfo == null) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Device not responding.");
```

Comment:
If the device is set off line here it looks like it will never be set online again. Because the first part of refresh checks if the device is online and if not just returns.

Reply:
See above. Now the authentication call happens before and puts the device back online, if authentication was successful.